### PR TITLE
Normalized return values and made vector classes use extendable types

### PR DESCRIPTION
**Pull request by LostInClams**
This commit tries to normalize the return type of all math functions so that all ToRef versions return the ref. It also changes so that functions that create new Objects try to do that with an extended type when possible.

More information can be found [here](https://forum.babylonjs.com/t/types-in-math-vector-ts/34514/13)

### CODE REVIEW

File: packages/dev/core/src/Maths/math.vector.ts [hunk 1]
```diff
@@ -73,7 +73,7 @@ export class Vector2 {
      * @param index defines the offset in source array
      * @returns the current Vector2
      */
-    public toArray(array: FloatArray, index: number = 0): Vector2 {
+    public toArray(array: FloatArray, index: number = 0): this {
```
Comment (BabylonJSGuide) [inline packages/dev/core/src/Maths/math.vector.ts:76]: Not an expert but I presume that this is not recognized as a type. You are still returning a Vector2 so it should remain as a Vector2 even though you are returning it in the form of this (a Vector2)

Same where you have changed other types to this rather than a named type.

File: packages/dev/core/src/Maths/math.vector.ts [hunk 2]
```diff
@@ -178,8 +178,8 @@ export class Vector2 {
      * @param otherVector defines the other vector
      * @returns a new Vector2
      */
-    public addVector3(otherVector: Vector3): Vector2 {
-        return new Vector2(this.x + otherVector.x, this.y + otherVector.y);
+    public addVector3(otherVector: Vector3): this {
+        return new (<any>this.constructor)(this.x + otherVector.x, this.y + otherVector.y);
```
Comment (RaananW) [inline packages/dev/core/src/Maths/math.vector.ts:182]: I like the generics approach, but I am not sure I am happy about `this.constructor` casted to `any` to get the types correctly.
I'll try playing around with it later, unless you have a suggestion with no `any` :-)

**Review by RaananW — COMMENTED** [review]
I like the changes a lot!
As written in the comment, I am simply not too happy about the `any` cast. I know it cannot be solved directly with generics, but I want to see if we find a more elegant solution.

File: packages/dev/core/src/Maths/math.vector.ts (hunk 2 above)
Comment (RaananW) [reply packages/dev/core/src/Maths/math.vector.ts:182]: How about something like this:

```javascript
export type Vector4Constructor<T extends Vector4> = new (x: number, y: number, z: number, w: number) => T;
//... now the vector 4 add function:
    /**
     * Returns a new Vector4 as the result of the addition of the current Vector4 and the given one.
     * @param otherVector the vector to add
     * @returns the resulting vector
     */
    public add(otherVector: DeepImmutable<Vector4>): this {
        return new (<Vector4Constructor<this>>this.constructor)(this.x + otherVector.x, this.y + otherVector.y, this.z + otherVector.z, this.w + otherVector.w);
    }
``

**Comment by RaananW** [comment]
Also note - tests are failing, so something else changed other than types.

**Comment by LostInClams** [comment]
Yes, I did change many places to return ref instead of this in ToRef functions, I felt that we agreed to start of from there since Sebavan and Bghgary thought it was on the edge case for a bug :)

Tried to run the test but haven't really contributed to the Babylon source before so not sure how to get the test results to figure out if the failing cases are related to the this vs ref return or something if I broke something else. Could you point me to some guide for the tests?

**Comment by RaananW** [comment]
> Yes, I did change many places to return ref instead of this in ToRef functions, I felt that we agreed to start of from there since Sebavan and Bghgary thought it was on the edge case for a bug :)
> 
> Tried to run the test but haven't really contributed to the Babylon source before so not sure how to get the test results to figure out if the failing cases are related to the this vs ref return or something if I broke something else. Could you point me to some guide for the tests?

Sure. Unit tests are executed using `npm run test:unit`. Visualization testss are executed using `npm run test:visualization`, but you also need the local babylon-server in the background, so run `npm run serve -w @tools/babylon-server` in a different terminal first

File: packages/dev/core/src/Maths/math.vector.ts (hunk 2 above)
Comment (LostInClams) [reply packages/dev/core/src/Maths/math.vector.ts:182]: @RaananW To have this work I tried to modify your proposal like this
`export type Vector2Constructor<T extends Vector2> = (new () => T) | ((new (x: number, y: number) => T));`
But this still gives me "Expected 2 arguments, but got 0." when I try to use it like this
`const result = new (<Vector2Constructor<T>>value1.constructor)();`
I guess one solution is to not use the default constructor or do you have a better idea? :)

File: what's new.md [hunk 3]
```diff
@@ -284,6 +284,7 @@
 
 - Faster scalar's WithinEpsilon with Math.abs ([nekochanoide](https://github.com/nekochanoide))
 - Added `decomposeToTransformNode` ([RaananW](https://github.com/RaananW))
+- Normalized types and return values in math.vector.ts ([LostInClams](https://github.com/LostInClams))
```
Comment (RaananW) [inline what's new.md:287]: please revert this, there is no need to change the what's new file :-)

File: packages/dev/core/src/Maths/math.vector.ts (hunk 2 above)
Comment (RaananW) [reply packages/dev/core/src/Maths/math.vector.ts:182]: the variables should be optional, similar to the original constructor, i.e. `x?: number, y?: number`

File: what's new.md (hunk 3 above)
Comment (LostInClams) [reply what's new.md]: Should I fix line 63 in contributing.md as well?

"- Make sure to add a line about your PR in the [what's new](https://github.com/BabylonJS/Babylon.js/blob/master/what's%20new.md)"

Comment (RaananW) [reply what's new.md]: Yep! thanks for that :-)

**Review by bghgary — COMMENTED** [review]
I'm not 100% sure what I suggested works, but if it does, it is maybe cleaner?

File: packages/dev/core/src/Maths/math.vector.ts [hunk 4]
```diff
@@ -12,6 +12,12 @@ import { EngineStore } from "../Engines/engineStore";
 
 type TransformNode = import("../Meshes/transformNode").TransformNode;
 
+export type Vector2Constructor<T extends Vector2> = new (x?: number, y?: number) => T;
+export type Vector3Constructor<T extends Vector3> = new (x?: number, y?: number, z?: number) => T;
+export type Vector4Constructor<T extends Vector4> = new (x?: number, y?: number, z?: number, w?: number) => T;
+export type QuaternionConstructor<T extends Quaternion> = new (x?: number, y?: number, z?: number, w?: number) => T;
+export type MatrixConstructor<T extends Matrix> = new () => T;
+
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/core/src/Maths/math.vector.ts:20]: that would be the same as casting everything to `any` . Not sure how much I like it. What is wrong with these types?

[the remark this answers is not in the document]
Comment (LostInClams) [reply packages/dev/core/src/Maths/math.vector.ts:20]: I also like the type we have now more, since it forces you to use the same constructor arguments as intended by each class.

File: packages/dev/core/src/Maths/math.vector.ts [hunk 5]
```diff
@@ -178,8 +184,8 @@ export class Vector2 {
      * @param otherVector defines the other vector
      * @returns a new Vector2
      */
-    public addVector3(otherVector: Vector3): Vector2 {
-        return new Vector2(this.x + otherVector.x, this.y + otherVector.y);
+    public addVector3(otherVector: Vector3): this {
+        return new (<Vector2Constructor<this>>this.constructor)(this.x + otherVector.x, this.y + otherVector.y);
```
[the remark this answers is not in the document]
Comment (LostInClams) [reply packages/dev/core/src/Maths/math.vector.ts:188]: Is there a reason you prefer `as T` casts over `<T>`?

[the remark this answers is not in the document]
Comment (bghgary) [reply packages/dev/core/src/Maths/math.vector.ts:188]: `<T>value` is the old, deprecated way of casting. https://stackoverflow.com/questions/46080409/what-is-the-difference-between-angle-brackets-and-as-in-typescript-2

File: packages/dev/core/src/Maths/math.vector.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (bghgary) [reply packages/dev/core/src/Maths/math.vector.ts:20]: Ugh, yeah, true. Mine won't enforce argument types. It stinks that there is no generic way to get the constructor type.

That said, I did find a way such that you don't have to duplicate the arguments of the constructor:

```typescript
export type Vector2Constructor<T extends Vector2> = new (...args: ConstructorParameters<typeof Vector2>) => T;
```

[the remark this answers is not in the document]
Comment (bghgary) [reply packages/dev/core/src/Maths/math.vector.ts:20]: I found a generic way, but it's probably too much.

```typescript
export type Constructable<T1 extends abstract new (...args: any) => any, T2> = new (...args: ConstructorParameters<T1>) => T2;

//...

const vector = new (this.constructor as Constructable<typeof Vector3, this>)(0, 0, 0);
```

File: packages/dev/core/src/Maths/math.vector.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (LostInClams) [reply packages/dev/core/src/Maths/math.vector.ts:188]: Ok, I’ll fix it but calling it deprecated seemed a bit strong. It’s also the way casting has been done in the past in the file :)

File: packages/dev/core/src/Maths/math.vector.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (LostInClams) [reply packages/dev/core/src/Maths/math.vector.ts:20]: While I find this cool, it feels like your other proposal is a lot more readable to me. 👍

File: packages/dev/core/src/Maths/math.vector.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (bghgary) [reply packages/dev/core/src/Maths/math.vector.ts:188]: That's fair. Maybe a bit strong wording there. 😄

**Review by bghgary — APPROVED** [review]

File: packages/dev/core/src/Maths/math.vector.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (LostInClams) [reply packages/dev/core/src/Maths/math.vector.ts]: Seems like the visualization tests don't like this change :/ And can't really find any information about their setup so not sure how I should go about debugging this..

**Review by RaananW — APPROVED** [review]
